integration_test/EmitVerilog/basic.mlir

```diff
@@ -1,12 +1,11 @@
 // REQUIRES: verilator
 // RUN: circt-translate %s -emit-verilog -verify-diagnostics > %t1.sv
-// RUN: verilator --cc --top-module main -sv %t1.sv --build --exe %S/../driver.cpp
-// RUN: ./obj_dir/Vmain --cycles 8 2>&1 | FileCheck %s
+// RUN: circt-rtl-sim.py verilator %t1.sv %S/../driver.cpp --simargs "--cycles 8" 2>&1 | FileCheck %s
```

Comment:
Instead of pushing the choice of simulator into the test, can you push it into circt-rtl-sim?  Rationale: There is nothing verilator specific about this test, that is just an implementation detail.  I'd prefer to see something like:

```
// REQUIRES: rtl-sim
// RUN: circt-translate %s -emit-verilog -verify-diagnostics > %t1.sv
// RUN: circt-rtl-sim.py --top-module top %t1.sv --cycles 8

or something like that.  the fact that verilator poops out a C++ file etc should all be handled within the script.  common flags like "--cycles 8" that can be supported by multiple simulators can then be handled by the script.

To put it another way, I'd love for this script to be the [`lli`](https://llvm.org/docs/CommandGuide/lli.html) of system verilog.

Reply:
That would be difficult and not terribly useful. Or very, very, extremely hard and very useful.

> Rationale: There is nothing verilator specific about this test, that is just an implementation detail.

Actually, there is. The `driver.cpp` program (which is part of the test) relies on several characteristics: the top level is specified as top, and there are `clk` and `rstn` pins which need to be driven appropriately. Testbenches which aren't written to target verilator would not expose any ports in their top module, so the produced SystemVerilog is intended for Verilator. They also don't need a C++ driver program.

This is a sort of contrived example as it is incredibly uninteresting. It just runs the produced SystemVerilog toggling the `clk` and `rstn` ports. Real tests would want to expose more I/O ports to test functionality. If using verilator, they'd have to write a custom C++ driver app (which could presumably control the number of cycles). If using a commercial simulator, the top level wouldn't expose any ports and it would drive and read internal wires via simulation constructs which verilator doesn't support. Would it be possible to automatically synthesize a C++ driver app with the same functionality? Yes, but said functionality isn't terribly interesting so it wouldn't be used much.

Bottom line: every single test today is written to target either verilator or a commercial simulator as they are _very_ different beasts. Should we do interesting things to advance the status quo? Absolutely, though that is out-of-scope for this script and would have to be thought through further. Should we create a 'simulation' dialect which can lower to either a C++ driver (for verilator) or commercial simulation constructs? Maybe. Should we require all testing be done via synthesizable SystemVerilog (so that it's supported by verilator and commercial sims)? Probably not.

I fully expect that for the foreseeable future that for anything interesting, it would have to target either Verilator or a commercial simulator.

If you look at the example I put in the description, you'll see that I did add footing for selecting a commercial simulator and using it. There your assumption that there is nothing simulator-specific applies. It wouldn't be too much harder to pass the default simulator to the script via an environment variable (to eliminate the %defaultSim) so I'll do that.

Reply:
Caveat: Running an `uninteresting` simulation (only `clk` and `rst`) becomes interesting when paired with Cosim to drive it from another process. So I'll probably add test driver shell generation (C++ or SV) at some point in the future to support ESI cosim simulations.

Reply:
Ok, I understand what you're saying here, but I still think you're missing my point.  :-)

We're likely to want many _simple_ tests for basic functionality that are not doing fancy things with clocks and reset logic.  We can have standardized ".cpp" file for verilator that works with "the one true way" and we can have a standardized module wrapper that gets thrown into vcs/questa or whatever.   We can just require all these style of tests to define their top level module with the same name like "top" - no need to get fancy here.

That said, this doesn't eliminate the utility of having a questa/verilator specific driver script for doing questa/verilator specific things.  It is just that simulator independent tests like this would have to be duplicated, and that will make me grumpy every time it happens :)

Investing in test infra like this pays off big time in the long run, because getting it wrong up front creates a lot of the "wrong thing" that has to be reworked later.

Reply:
I added an SystemVerilog driver which has the option to kill the simulation after N cycles, just like driver.cpp. This allowed me to run basic_run_test.sv in both verilator and questa without changing it. I'm not confident that this'll be useful for anything other than self-stimulating tests and uninteresting tests, but time will tell.

Still have some cleanup to do. Just pushing for the night.